README.md
# 🔎 Error Lens for Neovim

An enhanced visual diagnostic display for Neovim, inspired by the [Error Lens extension for Visual Studio Code](https://github.com/usernamehw/vscode-error-lens).

## 📸 Showcase

![Showcase of all diagnostic hints](media/error_lens_showcase.png "Showcase of all diagnostic hints")

## ✨ Features

- Visually supportive diagnostic messages
- Telescope Extension sorting by severity
- Customizable colors and appearance
- Auto-adjust colors based on the theme color
- Easy setup with all package managers
- Togglable
- Heavily inspired by the Error Lens plugin for Visual Studio Code

## 🚧 Roadmap

- [x] Telescope Extension
- [x] Togglable hints
- [x] Auto adjust colors based on theme
- [x] Custom colors

## ⚡️ Requirements

- Neovim >= 0.5
- Properly configured Neovim LSP client

## 📦 Installation

Install the plugin with your preferred package manager:

### packer

```lua

use {
    'chikko80/error-lens.nvim',
    requires = {'nvim-telescope/telescope.nvim'}
}

require("error-lens").setup(client, {
    -- your options go here
})
```

### lazy.nvim

```lua
{
    "chikko80/error-lens.nvim",
    event = "BufRead",
    dependencies = {
        "nvim-telescope/telescope.nvim"
    }
    opts = {
        -- your options go here
    },
}
```

## ⚙️ Configuration

error-lens.nvim comes with the following default options:

```lua
{
  -- this setting tries to auto adjust the colors
  -- based on the diagnostic-highlight groups and your
  -- theme background color with a color blender
  enabled = true,
  auto_adjust = {
    enable = false,
    theme_bg = nil, -- mandatory if enable true (e.g. #281478)
    step = 5, -- inc: colors should be brighter/darker
    total = 30 -- steps of blender
  },
  prefix = 4, -- distance code <-> diagnostic message
  -- default colors
  colors = {
    error_fg = "#FF6363", -- diagnostic font color
    error_bg = "#4B252C", -- diagnostic line color
    warn_fg = "#FA973A",
    warn_bg = "#403733",
    info_fg = "#5B38E8",
    info_bg = "#281478",
    hint_fg = "#25E64B",
    hint_bg = "#147828"
  }
}
```

## 🚀 Usage

After installing and configuring the plugin, it will automatically enhance the diagnostic display in Neovim.

To call the Telescope Error Lens `:ErrorLensTelescope`

To disable/toggle the plugin temporarily, use `:ErrorLensToggle`

## ⌨️ You just switch from VsCode and used OneDarkPro?

If you're switching from VSCode and are used to the beloved [OneDarkPro Theme](https://marketplace.visualstudio.com/items?itemName=zhuangtongfa.Material-theme),
you can use its equivalent for Neovim: [olimorris/onedarkpro.nvim](https://github.com/olimorris/onedarkpro.nvim).

The default colors are set up for this theme, and i've tried my best to match the colors (despite the lack of opacity in nvim).

## ⚠️ Disclaimer:

I recently switched from VsCode to Neovim and really missed that plugin. Therefore i tried to code it myself.
I am not really familiar with Lua, so if you find any bugs please raise an issue.
PRs and improvements or ideas are also welcome.

## 📚 Credits

This plugin is heavily inspired by the [Error Lens extension](https://github.com/usernamehw/vscode-error-lens) for Visual Studio Code.


CHANGELOG.md
## [1.0.2]

### Added
- Telescope Extension - Sort diagnostics by severity

## [1.0.1]

### Added
- Toggle with :ErrorLensToggle

### Changed
- Highlighting function to highlight the full line (Simplified version)

### Removed
- disable_vt - we disable virtual_text by default an restore the old setting on toggle


## [1.0.0]

### Added
- Added a new setup function without parameters to hook into LSP clients' diagnostics.
- Implemented a custom diagnostics handler in the setup function to update highlights.
- The custom diagnostics handler is now attached to Neovim's built-in LSP framework by modifying `vim.lsp.handlers["textDocument/publishDiagnostics"]`.

### Changed
- Simplified the plugin setup process by removing the need for specifying LSP clients as parameters.
- The custom diagnostics handler now works with all active LSP clients automatically.

### Removed
- Removed the old setup function that required specifying LSP clients as parameters.


lua/error-lens/init.lua
local config = require("error-lens.config")
local utils = require("error-lens.utils")
local highlight = require("error-lens.highlight")

local function setup(options)
	vim.cmd('command! ErrorLensToggle lua require("error-lens").toggle()')
	vim.cmd("command! ErrorLensTelescope lua require('error-lens.telescope').show_diagnostics()")

	config.setup(options)

	vim.diagnostic.handlers.error_lens = {
		---@param bufnr number
		---@param diagnostics table
		show = function(_, bufnr, diagnostics, _)
			highlight.update_highlights(bufnr, diagnostics)
		end,
		---@param bufnr number
		hide = function(_, bufnr)
			highlight.clear_highlights(bufnr)
		end,
	}
end

local function toggle()
	local value = not vim.diagnostic.config().error_lens
	config.options.enabled = value

	if value then
		-- enable error lens -> disable default virtual_text
		utils.set_virtual_text(false)
	else
		-- disable_error_lens -> restore user setting virtual_text
		utils.set_virtual_text(config.user_virtual_text)
	end

	vim.diagnostic.config({ error_lens = value })

	print("ErrorLens is now", config.options.enabled and "enabled" or "disabled")
end

return {
	setup = setup,
	toggle = toggle,
}


lua/error-lens/utils.lua
local function hexToRGB(hex)
	hex = hex:gsub("#", "")
	return tonumber("0x" .. hex:sub(1, 2)), tonumber("0x" .. hex:sub(3, 4)), tonumber("0x" .. hex:sub(5, 6))
end

function RGBToHex(r, g, b)
	return string.format("#%02X%02X%02X", r, g, b)
end

local function color_bend(hex1, hex2, step, total)
	local r1, g1, b1 = hexToRGB(hex1)
	local r2, g2, b2 = hexToRGB(hex2)
	local t = step / total

	local r = r1 + (r2 - r1) * t
	local g = g1 + (g2 - g1) * t
	local b = b1 + (b2 - b1) * t

	return RGBToHex(math.floor(r), math.floor(g), math.floor(b))
end

local function get_default_diagnostic_colors(theme_bg, hi_group, step, total)
	local color = vim.api.nvim_get_hl_by_name(hi_group, true)
	local fg = string.format("#%06x", color.foreground)
	local bg

	if color.background ~= nil then
		bg = string.format("#%06x", color.background)
	else
		bg = color_bend(theme_bg, fg, step, total)
	end

	return fg, bg
end

local function set_virtual_text(value)
	vim.diagnostic.config({
		virtual_text = value,
	})
end

return {
	get_default_diagnostic_colors = get_default_diagnostic_colors,
	set_virtual_text = set_virtual_text,
}


lua/error-lens/highlight.lua
local config = require("error-lens.config")

local function set_highlight(buf, diagnostic, fg, bg)
	vim.api.nvim_buf_add_highlight(buf, config.namespace, fg, diagnostic.lnum, 0, -1)
	vim.api.nvim_buf_set_extmark(buf, config.namespace, diagnostic.lnum, 0, {
		hl_mode = "combine",
		hl_eol = true,
		hl_group = bg,
		virt_text = {
			{ string.rep(" ", 5) .. diagnostic.message, fg },
		},
		priority = 4 - diagnostic.severity,
	})
end

local function clear_highlights(buf)
	vim.api.nvim_buf_clear_namespace(buf, config.namespace, 0, -1)
end

local function update_highlights(buf, diagnostics)
	clear_highlights(buf)

	for _, diagnostic in ipairs(diagnostics) do
		if diagnostic.severity == vim.diagnostic.severity.ERROR then
			set_highlight(buf, diagnostic, "ErrorLensErrorText", "ErrorLensErrorBg")
		elseif diagnostic.severity == vim.diagnostic.severity.WARN then
			set_highlight(buf, diagnostic, "ErrorLensWarnText", "ErrorLensWarnBg")
		elseif diagnostic.severity == vim.diagnostic.severity.INFO then
			set_highlight(buf, diagnostic, "ErrorLensInfoText", "ErrorLensInfoBg")
		elseif diagnostic.severity == vim.diagnostic.severity.HINT then
			set_highlight(buf, diagnostic, "ErrorLensHintText", "ErrorLensHintBg")
		end
	end
end

return {
	update_highlights = update_highlights,
	clear_highlights = clear_highlights,
}


lua/error-lens/telescope.lua
local pickers = require("telescope.pickers")
local finders = require("telescope.finders")
local actions = require("telescope.actions")
local conf = require("telescope.config").values
local entry_display = require("telescope.pickers.entry_display")
local previewers = require("telescope.previewers")

local severity_labels = {
	[0] = "Other",
	[1] = "Error",
	[2] = "Warning",
	[3] = "Info",
	[4] = "Hint",
}

-- code from trouble.nvim
-- returns a hl or sign label for the givin severity and type
-- correctly handles new names introduced in vim.diagnostic
local function get_severity_label(severity, type)
	local label = severity
	local prefix = "LspDiagnostics" .. (type or "Default")

	if vim.diagnostic then
		prefix = type and ("Diagnostic" .. type) or "Diagnostic"
		label = ({
			Warning = "Warn",
			Information = "Info",
		})[severity] or severity
	end

	return prefix .. label
end

-- Extract information from diagnostics to display in telescope
local function preprocess_diagnostics(diagnostics)
	local items = {}
	for _, diagnostic in ipairs(diagnostics) do
		local bufnr = diagnostic.bufnr
		local filepath = vim.api.nvim_buf_get_name(bufnr)
		local display_filename = filepath:gsub(vim.loop.cwd() .. "/", "")
		local row, col = diagnostic.lnum, diagnostic.col

		local sign_label = severity_labels[diagnostic.severity]
		local sign_info = vim.fn.sign_getdefined(get_severity_label(sign_label, "Sign"))[1]
		local sign = vim.trim(sign_info.text)
		local hl_group = vim.trim(sign_info.texthl)

		local entry = {
			value = {
				bufnr = bufnr,
				lnum = row,
				col = col,
				message = vim.trim(diagnostic.message:gsub("[\n]", "")),
				path = display_filename,
				sign = sign,
				hl_group = hl_group,
				severity = diagnostic.severity, -- Add this line
			},
			ordinal = string.format("%s %d:%d | %s | %s", sign, row + 1, col + 1, diagnostic.message, display_filename),
		}

		table.insert(items, entry)
	end
	return items
end

-- Copied and modified from telescope.nvim
local function entry_maker(entry)
	local value = entry.value
	local bufnr = value.bufnr
	local row, col = value.lnum, value.col
	local display_line = string.format("%s %4d:%2d", value.sign, row + 1, col + 1)

	local display_items = {
		{ width = 10 },
		{ width = 60 },
		{ remaining = true },
	}
	local displayer = entry_display.create({
		separator = "▏",
		items = display_items,
	})

	return {
		display = function(entry)
			return displayer({
				{ display_line, value.hl_group },
				value.message,
				value.path,
			})
		end,
		ordinal = entry.ordinal,
		bufnr = bufnr,
		lnum = row + 1,
		col = col + 1,
		start = col,
		finish = col + #value.message,
		path = value.path,
	}
end

local function show_diagnostics(opts)
	opts = opts or {}

	local diagnostics = vim.diagnostic.get(nil)

	local items = preprocess_diagnostics(diagnostics)

	-- sort diagnostics by severity
	table.sort(items, function(a, b)
		return a.value.severity < b.value.severity
	end)

	pickers
		.new(opts, {
			prompt_title = "Error Lens Diagnostics",
			finder = finders.new_table({
				results = items,
				entry_maker = entry_maker,
			}),
			sorter = conf.generic_sorter(opts),
			attach_mappings = function(_, map)
				map("i", "<CR>", actions.select_default)
				map("n", "<CR>", actions.select_default)
				return true
			end,
			previewer = previewers.vim_buffer_vimgrep.new(opts), -- Add this line
		})
		:find()
end

return {
	show_diagnostics = show_diagnostics,
}


lua/error-lens/config.lua
local utils = require("error-lens.utils")

local M = {}

M.namespace = vim.api.nvim_create_namespace("error_lens_highlights")

local vim_error = "DiagnosticError"
local vim_warn = "DiagnosticWarn"
local vim_info = "DiagnosticInfo"
local vim_hint = "DiagnosticHint"

local default_options = {
	enabled = true,
	auto_adjust = {
		enable = false,
		theme_bg = nil,
		step = 5,
		total = 30,
	},
	prefix = 4,
	colors = {
		error_fg = "#FF6363",
		error_bg = "#4B252C",
		warn_fg = "#FA973A",
		warn_bg = "#403733",
		info_fg = "#387EFF",
		info_bg = "#20355A",
		hint_fg = "#16C53B",
		hint_bg = "#254435",
	},
}

function M.setup(options)
	-- Merge user options with default options
	M.options = vim.tbl_deep_extend("force", {}, default_options, options or {})

	-- If auto_adjust is enabled
	if M.options.auto_adjust.enable then
		if M.options.auto_adjust.theme_bg == nil then
			error("Error: 'theme_bg' is mandatory when 'auto_adjust' is enabled.")
		else
			local step = M.options.auto_adjust.step
			local total = M.options.auto_adjust.total

			-- Get adjusted colors based on the theme_bg color
			M.options.colors.error_fg, M.options.colors.error_bg =
				utils.get_default_diagnostic_colors(M.options.auto_adjust.theme_bg, vim_error, step, total)
			M.options.colors.warn_fg, M.options.colors.warn_bg =
				utils.get_default_diagnostic_colors(M.options.auto_adjust.theme_bg, vim_warn, step, total)
			M.options.colors.info_fg, M.options.colors.info_bg =
				utils.get_default_diagnostic_colors(M.options.auto_adjust.theme_bg, vim_info, step, total)
			M.options.colors.hint_fg, M.options.colors.hint_bg =
				utils.get_default_diagnostic_colors(M.options.auto_adjust.theme_bg, vim_hint, step, total)
		end
	end

	-- Set highlights using the M.options.colors values
	vim.api.nvim_set_hl(M.namespace, "ErrorLensErrorText", { fg = nil, bg = M.options.colors.error_bg })
	vim.api.nvim_set_hl(
		M.namespace,
		"ErrorLensErrorBg",
		{ fg = M.options.colors.error_fg, bg = M.options.colors.error_bg }
	)

	vim.api.nvim_set_hl(M.namespace, "ErrorLensWarnText", { fg = nil, bg = M.options.colors.warn_bg })
	vim.api.nvim_set_hl(
		M.namespace,
		"ErrorLensWarnBg",
		{ fg = M.options.colors.warn_fg, bg = M.options.colors.warn_bg }
	)

	vim.api.nvim_set_hl(M.namespace, "ErrorLensInfoText", { fg = nil, bg = M.options.colors.info_bg })
	vim.api.nvim_set_hl(
		M.namespace,
		"ErrorLensInfoBg",
		{ fg = M.options.colors.info_fg, bg = M.options.colors.info_bg }
	)

	vim.api.nvim_set_hl(M.namespace, "ErrorLensHintText", { fg = nil, bg = M.options.colors.hint_bg })
	vim.api.nvim_set_hl(
		M.namespace,
		"ErrorLensHintBg",
		{ fg = M.options.colors.hint_fg, bg = M.options.colors.hint_bg }
	)

	vim.api.nvim_set_hl_ns(M.namespace)

	-- Save the user's original virtual_text setting
	M.user_virtual_text = vim.diagnostic.config().virtual_text

	utils.set_virtual_text(false)

	if M.options.enabled then
		vim.diagnostic.config({ error_lens = true })
	else
		vim.diagnostic.config({ error_lens = false })
	end
end

return M
